src/lib/core/immer-adapter/common/immutable-state-context.ts

```diff
@@ -39,4 +46,42 @@ export class ImmutableStateContext<T extends object> implements StateContext<T>
   public dispatch(actions: any | any[]): Observable<void> {
     return this.ctx.dispatch(actions);
   }
+
+  private clone<T = any>(obj: T): T {
+    let copy: any;
+
+    // Handle the 3 simple types, and null or undefined
+    if (null == obj || 'object' !== typeof obj) {
+      return obj;
+    }
+
+    // Handle Date
+    if (obj instanceof Date) {
+      copy = new Date();
+      copy.setTime(obj.getTime());
+      return copy as T;
+    }
+
+    // Handle Array
+    if (obj instanceof Array) {
+      copy = [];
+      for (let i = 0, len = obj.length; i < len; i++) {
+        copy[i] = this.clone(obj[i]);
+      }
+      return copy;
+    }
+
+    // Handle Object
+    if (obj instanceof Object) {
+      copy = {};
+      for (const attr in obj) {
```

Comment:
Does `for key in obj` look more semantic? I guess `attr` is a little bit confusing